### ISSUE TITLE
TigerVNC 1.6.0 on Windows 10 running as a Service is unbearably slow and does not work in user mode.

### ISSUE THREAD

**Issue by didibus** [TigerVNC/tigervnc#277]
I'm trying to get TigerVNC server 1.6.0 running on Windows 10. When I register it as a service, it can connect, but it runs really slow on a local network. Worse then any VNC I've ever tried.

If I unregister the service, and try running the server manually, then it doesn't work at all. I'll see the client having established a connection, but as soon as the client finishes typing his password, he just get disconnected.

**Comment by CendioOssman** [comment]
The poor performance is a known issue unfortunately. It has something to do with the hooks not working properly when WinVNC and the application are running as different users. If you switch to polling mode then it usually works better.

For the user mode, try adding `-Log *:file:100` to the command line. You'll get a log file in `c:\temp` at that point.

**Comment by calestyo** [comment]
shouldn't it be easy to trace that down, since it worked perfectly well in 1.5?

**Comment by CendioOssman** [comment]
I wasn't aware it was a regression on our side. That helps, thanks.
But the major issue is that we do not have anyone spending time looking at WinVNC.

**Comment by calestyo** [comment]
Interestingly... downgrading doesn't seem to fix it back...

**Comment by eddieray** [comment]
I'm not sure of the cause, but I used to use TightVNC (server) on Windows 8 and it was almost as good as RDP in terms of performance (I'm sure TigerVNC was even better - but I just found TigerVNC).  Ever since I upgraded to Windows 8.1, and now Windows 10, it has become unusably slow.  I tried RealVNC, and while that is a tiny bit faster, it's still too slow to be of use to me.

**Comment by HeWhoWatches** [comment]
Just found this bug report and wanted to mention that we also noticed the slowness when installing 1.6.0 on a Windows 2012 server at the end of last year - we resolved the issue by reverting back to the 1.5.0 release. With regards the comment from calestyo regarding downgrading not fixing the problem, I seem to recall at one point there may have been an issue with the services not being properly removed during uninstallation. As the service names between 1.5.0 and 1.6.0 changed it was possible to end up with two services installed (I think - it's going back 6 months):
1.5.0 - Service Name: WinVNC4, Display Name: VNC Server Version 4
1.6.0 - Service Name: TigerVNC Server, Display Name: Provides remote access to this machine via the VNC/RFB protocol.

With regards the comment from 8th April regarding not having anyone looking at WinVNC - is this still the case?

**Comment by CendioOssman** [comment]
> With regards the comment from 8th April regarding not having anyone looking at WinVNC - is this still the case?

Yes. TigerVNC is still pretty much just Cendio and @bphinz. Neither of us has any interest in the Windows server. So someone who actually uses the thing would be very welcome in the project.

**Comment by kriston** [comment]
Well, that's too bad.  TigerVNC works perfectly on Unix servers with Windows clients so I'm grateful to you for that.
For Windows servers, UltraVNC worked well until the 1.2.1.x release which disconnects clients randomly.  I think it has a timeout problem since it disconnects any VNC client, whether TigerVNC, TightVNC, RealVNC, or UltraVNC, after random periods of idle activity.

**Comment by jhaohai** [comment]
Same issue hear. And I change the capture method from hook to poll make it slightly faster

**Comment by falconws** [comment]
When fix this critical bug ?
Can't use TigerVNC on Windows with this bug.

**Comment by CendioOssman** [comment]
I'm afraid the situation is still the same, noone is currently working on WinVNC.

You could try adding bounties to this issue and see if it attracts some attention from a new developer.

**Comment by stefanLeo** [comment]
Hi 

Regarding this WinVNC work I have a few question. Maybe you can answer them:
- Does this have anything to do with RealVNC and their rebranding to Connect?
- I just saw that the winvnc folder has the RealVNC headers in it - is/was this code somewhere available directly from RealVNC?
- Did RealVNC stop the open source implementation?
- Did you just take this code as the baseline and improve it?
- When did you stop supporting/implementation on the WinVNC part?
- Do others (i.e. UltraVNC) also use the same WIN code base?
- What is your general opinion about the future of VNC on Windows? Is there an open source alternative (other than RDP)?

Thank you! & BR
Stefan

**Comment by CendioOssman** [comment]
> Regarding this WinVNC work I have a few question. Maybe you can answer them:
> 
>  *  Does this have anything to do with RealVNC and their rebranding to Connect?

Nope. RealVNC went proprietary (protocol and implementation) many years ago and as such they are not really a part of the VNC community.

> *  I just saw that the winvnc folder has the RealVNC headers in it - is/was this code somewhere available directly from RealVNC?

TigerVNC is originally based on RealVNC 4.0, the last open source version of RealVNC.

>  * Did RealVNC stop the open source implementation?

Yes. Both code and protocol went secret many years ago.

>   * Did you just take this code as the baseline and improve it?

Yup.

>  *  When did you stop supporting/implementation on the WinVNC part?

We never really started. To a large extent it's been untouched since RealVNC wrote it.

>  * Do others (i.e. UltraVNC) also use the same WIN code base?

Some might, but I think UltraVNC wrote their code from scratch (or possibly based it on the older RealVNC 3 code base).

> *  What is your general opinion about the future of VNC on Windows? Is there an open source alternative (other than RDP)?

I'm afraid I don't keep much track of this, but AFAIK UltraVNC is the most popular implementation for Windows.

**Comment by mhertz** [comment]
Please forgive me for "off-topic", but i write here because there where two previous posts here with same type question.

I just found tigervnc and love it, thanks! I have it on linux, as server and client, but for when i need to give remote-support to a family windows-user, then would I be better of installing some other server than tigervnc because it's not really looked into and "for a large extent it's been untouched since RealVNC wrote it" ? I was thinking of installing tightvnc server on the win machine, as that is rewritten for windows and "looked into" by the devs, but I then am uncertain if I loose some optimization from using tigervnc only as client but not server, e.g. when going from tigervnc-client(linux) to tigervnc-server(windows) it states "enabling continous updates", whereas that line is missing when using tigervnc-client(linux) to tightvnc server(windows).

Also, I can't set screen polling interval from tigervnc-server(i disabled hooks, as slow), so what is it set at as default i.e. tightvnc is 1000, but can be set down to 30 min.

Thanks alot in advance!

**Comment by CendioOssman** [comment]
> Also, does tigervnc's windows server support the desktop duplication api of windows 8 and newer windows versions? It's supposed to give much faster screen-updates and is implemented in e.g. tightvnc and ultravnc.

No, that is not something we've looked at. However it seems to fit what we need quite nicely. But the main problem is still finding a developer willing to work on it.

**Comment by mhertz** [comment]
Thanks alot CendioOssman for the reply, much appreciated! :) Of course I understand that.

**Comment by count49** [comment]
Hey there. Here's a fun fact: the hooking/polling method isn't just obsolete, it's a veritable Ruby Goldberg machine that should've been retired in favor of Mirror Drivers feature some *eighteen* years ago, when Windows 2000 was introduced. Thus I believe I'll spend this week bringing Winvnc into the 21st millenium by writing new Mirror Drivers and Duplication SDisplayCore implementations, possibly lifting bits and pieces from Freerdp's Windows server implementation. Cheers!

**Comment by kriston** [comment]
Thanks.  If it is possible to preserve the brightness control on laptops
that would be even better.  So far, every mirror driver I've used
(TigerVNC, UltraVNC, LogMeIn, SplashTop) kills the brightness control on
most laptops. 

Kriston 

On 2018-07-09 16:00, count49 wrote:

> Hey there. Here's a fun fact: the hooking/polling method isn't just obsolete, it's a veritable Ruby Goldberg machine that should've been retired in favor of Mirror Drivers feature some _eighteen_ years ago, when Windows 2000 was introduced. Thus I believe I'll spend this week bringing Winvnc into the 21st millenium by writing new Mirror Drivers and Duplication SDisplayCore implementations, possibly lifting bits and pieces from Freerdp's Windows server implementation. Cheers! 
> 
> --
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub [1], or mute the thread [2].
 

Links:
------
[1]
https://github.com/TigerVNC/tigervnc/issues/277#issuecomment-403601656
[2]

**Comment by the-Arioch** [comment]
@count49 - any news?

Also, both TightVNC (which is parent of TigerVNC) and UltraVNC (which uses TurboVNC code like TigerVNC) have mirror drivers on their sites. But different ones, at least in names. UVNC talks about Windows versions ending with Win7, TightVNC just do not list any.
 
So, if anyone is in the know, 
1) are those drivers different projects, or one if fork of the other?
2) which of them is "better" - more up to date, more stable, etc
3) can TigerVNC just use one of those drivers instead of righting their own (count49?)

I mean, if there is "one VNC community" but it has few Windows-interested developers - why dissipate forces? All Windows VNC servers are to some degree forks of one another. Then their requirements for mirror driver should be very similar, or maybe identical. Why not making One True Windows VNC driver, when you all clearly do not have enough hands to make half-dozen of per-fork drivers?

**Comment by the-Arioch** [comment]
Actually, TightVNC seem to use this one - http://www.demoforge.com/dfmirage.htm - the company tried to make commercial "Screencast for tutorials" software out of TightVNC and probably implementing that mirror drivers functionality for it... But seems they are defunct since 2010

**Comment by the-Arioch** [comment]
uVNC says that starting with Windows 8 there is no need for mirror drivers: http://www.uvnc.com/downloads/screenrecorder/86-screenrecorder.html
This may be the reasons those both drivers frozen at Win7. 
This however may be the reason in slow work of services, as starting with Vista letting services "interact with desktop" is more and more suppressed for security

P.S. Those DemoForge files, I wonder what they are - there is a clear cut-off RFB ServerInit packet at the beginning, then some fixed bytes, then it goes strange (encrypted?)

**Comment by grnassar** [comment]
> The poor performance is a known issue unfortunately. It has something to do with the hooks not working properly when WinVNC and the application are running as different users. If you switch to polling mode then it usually works better.

Now that there's a bounty I imagine a workaround isn't sufficient to consider the ticket closed -- but switching the server from hooks to polling mode made an ENORMOUS difference.  The improvement was dramatic.  Given that work on the Windows version is necessarily limited, is there any way at least that the Windows installation can be made to default to polling mode instead of hooks?  This would seem to be the "low-hanging fruit" ideal for-now solution to the problem.

**Comment by CendioOssman** [comment]
That seems reasonable, yes. I've changed the default in d06906db59a6aa6434a5dd602b8d3c987d291ca8.

**Comment by michaellleslie** [comment]
Is this still an issue?

**Comment by CendioOssman** [comment]
Likely. No one has worked on it at least.

**Comment by grnassar** [comment]
> Is this still an issue?

Probably still an issue if you choose to use hooks instead of polling, but I'm not sure why anyone would choose that.  Default Windows install configures the server to use polling now rather than hooks, and the speed on Windows is now far better and very usable.  For all practical purposes I can think of that has been a successful "fix."
